### PR TITLE
Add container_run_and_commit_layer rule

### PR DESCRIPTION
**Pull request by jamiees2**
The `container_run_and_commit` rule outputs an image, but sometimes all you want is the layer of changes to the root image. 

We had a use-case that wasn't properly covered by `container_commit_and_extract` where we would like to obtain all the changes to the image file system, but don't have a fixed set of files that are modified. We'd like to be able to reuse the layer, which motivated the creation of this new rule. 

We're also planning on replacing all our usages of `container_run_and_commit` with this rule. Internally, we mostly use it for creating custom base images, but using this rule we can get smaller outputs for that target, allowing us to better utilize our build cache.

### CODE REVIEW

**Comment by jamiees2** [comment]
@googlebot I signed it!

File: docker/util/run.bzl [hunk 1]
```diff
@@ -277,6 +285,186 @@ commit = struct(
...
+        },
+        is_executable = True,
+    )
+
+    docker_env = [
+        "{}={}".format(
+            ctx.expand_make_variables("env", key, {}),
+            ctx.expand_make_variables("env", value, {}),
+        )
+        for key, value in env
```
Comment (huan086) [inline docker/util/run.bzl:345]: When I specify the `env` argument, I get `type 'string' is not iterable`. This line should probably be `for key, value in env.items()`

Comment (jamiees2) [reply docker/util/run.bzl:345]: oh interesting, I wonder how it worked for me 🤔 Thanks!
```suggestion
        for key, value in env.items()
```

File: docker/util/README.md [hunk 2]
```diff
@@ -137,3 +138,79 @@ bazel-out directory.
...
+          A list of commands to run (sequentially) in the container.
+        </p>
+      </td>
+    </tr>
+    <tr id="container_run_and_commit_layer-docker_run_flags">
+      <td><code>docker_run_flags</code></td>
+      <td>
+        List of strings; optional
+        <p>
+          Extra flags to pass to the docker run command.
```
Comment (huan086) [inline docker/util/README.md:183]: I was struggling to get this rule to run. Kept getting strange errors of `Error: Unable to access jarfile sh`, until I realize that the entrypoint of `gcr.io/distroless/java-debian10` is `["/usr/bin/java", "-jar"]` 🤦‍♀️.

Would be good to remind users to clear the entrypoint here and give an example

**Comment by huan086** [comment]
I've got the rule running, this rule is well and good.

Found a strange behaviour with container_image.

```py
container_image(name = "to_run" ...)
container_run_and_commit_layer(name = "install", image = ":to_run.tar", ...)
container_image(name = "final", layers = [":install"], ...)
```

When I run the `final` target, I get a Docker image with both contents from `to_run` and `final` and the `manifest.json` in the image tar is

```json
[
    {
        "Config": "redacted1.json",
        "RepoTags": [
            "bazel/stuff:to_run"
        ],
        ...
    },
    {
        "Config": "redacted2.json",
        "RepoTags": [
            "bazel/staff:final"
        ],
        ...
    }
]
```

If I skip the container_run_and_commit_layer rule, it similarly results in the both contents in the Docker image.

```py
container_image(name = "to_run" ...)
container_image(name = "final", tars = [":to_run-layer.tar"], ...)
```


Do you face the same problem? If not, how does your build look like?

**Comment by jamiees2** [comment]
Hey @alex1545, @smukherj1, is there a chance of getting this reviewed sometime soon? :)

File: docker/util/run.bzl [hunk 3]
```diff
@@ -277,6 +285,186 @@ commit = struct(
...
+        commands = None,
+        docker_run_flags = None,
+        env = None,
+        compression = None,
+        compression_options = None,
+        output_layer_tar = None):
+    """Implementation for the container_run_and_commit_layer rule.
+
+    This rule runs a set of commands in a given image, waits for the commands
+    to finish, and then commits the container to a new image.
```
Comment (alex1545) [inline docker/util/run.bzl:301]: Please update the description accordingly.

**Review by alex1545 — COMMENTED** [review]
Thank you very much for your contribution and sorry for the delayed reply on this.
Can you also please add a simple test to exercise this new rule? Maybe a test that runs a command to add a new file and then confirming that the expected layer tar is generated containing the new file.

**Comment by jamiees2** [comment]
Hey @alex1545, I've addressed the PR review comments and added a test to verify that the rule works correctly. Is this good the be merged?

Sorry about the CLA mess, I made a git mistake when merging the upstream master into my branch, which I force reverted.

**Comment by alex1545** [comment]
Thank you for addressing the comments.
Can you please add your new test targets to the buildkite CI [here](https://github.com/bazelbuild/rules_docker/blob/d812ccfa59ae537b4fa27579a63a3d291e168cd5/.bazelci/presubmit.yml#L52), [here](https://github.com/bazelbuild/rules_docker/blob/d812ccfa59ae537b4fa27579a63a3d291e168cd5/.bazelci/presubmit.yml#L117) and [here](https://github.com/bazelbuild/rules_docker/blob/d812ccfa59ae537b4fa27579a63a3d291e168cd5/.bazelci/presubmit.yml#L189).
Also please take a look as to why the test fails on RBE with this error:


```
ERROR: /workdir/tests/docker/util/BUILD:110:10: in _rule_test_rule rule //tests/docker/util:test_container_commit_layer_rule_impl:
--
  | Traceback (most recent call last):
  | File "/workdir/tests/docker/util/BUILD", line 110, column 10, in _rule_test_rule(name = 'test_container_commit_layer_rule_impl')
  | rule_test(
  | File "/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/bazel_tools/tools/build_rules/test_rules.bzl", line 296, column 17, in _rule_test_rule_impl
  | fail("rule %s generates %s not %s" %
  | Error in fail: rule //tests/docker/util:test_container_commit_layer generates ["test_container_commit_layer-layer.tar"] not ["test_container_commit.build", "test_container_commit_commit-layer.tar"]
```

**Comment by jamiees2** [comment]
Thanks, sorry about that. I've fixed the test and added the two new test targets to the CI.

**Comment by alex1545** [comment]
Can you now please look at why this fails on Darwin with this error:

```
ERROR: /Users/buildkite/builds/bk-imacpro-8/bazel/rules-docker-docker/tests/docker/util/BUILD:100:31: Action tests/docker/util/test_container_commit_layer-layer.tar failed (Exit 1): test_container_commit_layer.build failed: error executing command
--
  | (cd /private/var/tmp/_bazel_buildkite/5e65cabdc7f99fb85376aa4a05c66e92/sandbox/darwin-sandbox/1034/execroot/io_bazel_rules_docker && \
  | exec env - \
  | PATH=/Users/buildkite/Library/Caches/bazelisk/downloads/bazelbuild/bazel-3.5.0-darwin-x86_64/bin:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin \
  | bazel-out/darwin-fastbuild/bin/tests/docker/util/test_container_commit_layer.build)
  | Execution platform: @local_config_platform//:host
  |  
  | Use --sandbox_debug to see verbose messages from the sandbox test_container_commit_layer.build failed: error executing command
  | (cd /private/var/tmp/_bazel_buildkite/5e65cabdc7f99fb85376aa4a05c66e92/sandbox/darwin-sandbox/1034/execroot/io_bazel_rules_docker && \
  | exec env - \
  | PATH=/Users/buildkite/Library/Caches/bazelisk/downloads/bazelbuild/bazel-3.5.0-darwin-x86_64/bin:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin \
  | bazel-out/darwin-fastbuild/bin/tests/docker/util/test_container_commit_layer.build)
  | Execution platform: @local_config_platform//:host
  |  
  | Use --sandbox_debug to see verbose messages from the sandbox
  | + source bazel-out/darwin-fastbuild/bin/tests/docker/util/image_util.sh
  | + DOCKER=/usr/bin/docker
  | + DOCKER_FLAGS=
  | + [[ -z /usr/bin/docker ]]
  | ++ bazel-out/host/bin/contrib/extract_image_id bazel-out/darwin-fastbuild/bin/external/debian_base/image/image.tar
  | + image_id=6786f6a45544215e9e4411db2318c7f561009b38fcf7c88f2f5f656256ba72f8
  | + /usr/bin/docker load -i bazel-out/darwin-fastbuild/bin/external/debian_base/image/image.tar
  | bazel-out/darwin-fastbuild/bin/tests/docker/util/test_container_commit_layer.build: line 19: /usr/bin/docker: No such file or directory
  | (16:53:46) ERROR: /Users/buildkite/builds/bk-imacpro-8/bazel/rules-docker-docker/tests/docker/util/BUILD:124:15 Action tests/docker/util/test_container_commit_layer-layer.tar failed (Exit 1): test_container_commit_layer.build failed: error executing command
  | (cd /private/var/tmp/_bazel_buildkite/5e65cabdc7f99fb85376aa4a05c66e92/sandbox/darwin-sandbox/1034/execroot/io_bazel_rules_docker && \
  | exec env - \
  | PATH=/Users/buildkite/Library/Caches/bazelisk/downloads/bazelbuild/bazel-3.5.0-darwin-x86_64/bin:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin \
  | bazel-out/darwin-fastbuild/bin/tests/docker/util/test_container_commit_layer.build)
  | Execution platform: @local_config_platform//:host
```

**Comment by jamiees2** [comment]
Oh I ran into this too when running the test locally, for me it was because of this line: https://github.com/bazelbuild/rules_docker/blob/master/WORKSPACE#L24

and my docker was at `/usr/local/bin/docker`. I changed that line temporarily so that I could run the test

**Comment by jamiees2** [comment]
Ah I see, we also don't run the analogous test for `container_run_and_commit` -  `//tests/docker/util:test_container_commit_metadata` either anywhere. For now, I'll remove it from OSX since it passed on other platforms, but lmk if you want it removed from the others as well.

**Review by alex1545 — APPROVED** [review]

**Comment by huan086** [comment]
I upgraded to [docker-desktop 2.4.2.0-edge](https://chocolatey.org/packages/docker-desktop/2.4.2.0-edge) and this rule broke with the error

```
ERROR: /bazel/elasticsearch/BUILD.bazel:183:31: Action elasticsearch/install_s3_plugin-layer.tar failed (Exit 125) install_s3_plugin.build failed: error executing command bazel-out/k8-fastbuild/bin/elasticsearch/install_s3_plugin.build

Use --sandbox_debug to see verbose messages from the sandbox
+ source bazel-out/k8-fastbuild/bin/elasticsearch/image_util.sh
+ DOCKER=/usr/bin/docker
+ DOCKER_FLAGS=
+ [[ -z /usr/bin/docker ]]
++ bazel-out/host/bin/contrib/extract_image_id bazel-out/k8-fastbuild/bin/elasticsearch/for_plugin_install.tar
+ image_id=50beeeb4436cfc02b264d10dca5bbc6965d7af83e14ccd1feb54306812ea0db4
+ /usr/bin/docker load -i bazel-out/k8-fastbuild/bin/elasticsearch/for_plugin_install.tar
++ /usr/bin/docker run -d --env-file bazel-out/k8-fastbuild/bin/elasticsearch/install_s3_plugin.env 50beeeb4436cfc02b264d10dca5bbc6965d7af83e14ccd1feb54306812ea0db4 sh -c '/usr/share/elasticsearch/bin/elasticsearch-plugin install --batch repository-s3'
docker: Error response from daemon: invalid repository name (50beeeb4436cfc02b264d10dca5bbc6965d7af83e14ccd1feb54306812ea0db4), cannot specify 64-byte hexadecimal strings.
+ id=fea7686c274c8b7bc931589967a344c2dadf9fde3e5e5d95b4761e7c99485269
Loaded image: pbrain/elasticsearch:for_plugin_install
```

I have to fix it by adding `sha256:` in `commit_layer.sh.tpl` line 21

```
id=$($DOCKER $DOCKER_FLAGS run -d %{docker_run_flags} --env-file %{env_file_path} sha256:"$image_id" %{commands})
```

**Comment by aryeh-looker** [comment]
Is there a way to log what's happening while running?

**Pull request merged**